### PR TITLE
2.x: Add Observable.rangeLong & Flowable.rangeLong

### LINKED ISSUE #4683: 2.x: Observable.range(long start, long count)
I'd like to have a range of long values as an Observable however only a range method with ints do exist. Is there a reason for this?

The only way I could imagine doing this now would be by using `intervalRange` with a low period however that does not seem right.

### PR DESCRIPTION
**Pull request by vanniktech**
Will fix #4683
- adds Observable.rangeLong()
- adds Flowable.rangeLong()

both implementations are copied same for the unit tests.

What should we do about the range overflow check? We could use BigDecimal to check that.

### CODE REVIEW

File: src/main/java/io/reactivex/Flowable.java [hunk 1]
```diff
@@ -3456,6 +3456,49 @@ public static int bufferSize() {
...
+
+        if (count == 0) {
+            return empty();
+        }
+
+        if (count == 1) {
+            return just(start);
+        }
+
+        if (start + (count - 1) > Long.MAX_VALUE) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/Flowable.java]: That is always false. You should check if start + count - 1 flips to negative.

File: src/main/java/io/reactivex/internal/operators/observable/ObservableRangeLong.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        void run() {
+            if (fused) {
+                return;
+            }
+            Observer<? super Long> actual = this.actual;
+            long e = end;
+            for (long i = index; i != e && get() == 0; i++) {
+                actual.onNext((long)i);
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/observable/ObservableRangeLong.java]: unnecessary cast to long

File: src/main/java/io/reactivex/Observable.java [hunk 3]
```diff
@@ -2993,6 +2993,46 @@ public static int bufferSize() {
...
+
+        if (count == 0) {
+            return empty();
+        }
+
+        if (count == 1) {
+            return just(start);
+        }
+
+        if (start + (count - 1) > Long.MAX_VALUE) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/Observable.java]: Same overflow check here as in Flowable.

**Review by akarnokd — CHANGES_REQUESTED** [review]
Please also make sure the coverage of Flowable and Observable itself remains 100%

File: src/main/java/io/reactivex/Flowable.java (hunk 1 above)
Comment (vanniktech) [reply src/main/java/io/reactivex/Flowable.java]: I know that's why I wrote `What should we do about the range overflow check? We could use BigDecimal to check that.`

File: src/main/java/io/reactivex/Observable.java (hunk 3 above)
Comment (vanniktech) [reply src/main/java/io/reactivex/Observable.java]: Same as above

File: src/main/java/io/reactivex/Flowable.java (hunk 1 above)
Comment (JakeWharton) [reply src/main/java/io/reactivex/Flowable.java]: You can do `Long.MAX_VALUE - count > start` or check just if it overflows negative like @akarnokd said.

Comment (akarnokd) [reply src/main/java/io/reactivex/Flowable.java]: See `intervalRange`:

``` java
        long end = start + (count - 1);
        if (start > 0 && end < 0) {
            throw new IllegalArgumentException("Overflow! start + count is bigger than Long.MAX_VALUE");
        }
```

Comment (vanniktech) [reply src/main/java/io/reactivex/Flowable.java]: Oh cool didn't know that. Will change it thanks.

**Review by akarnokd — APPROVED** [review]

**Pull request merged**